Do not clean directories directly under the application root with `Rails::BacktraceCleaner`

alpaca-tc opened the pull request:
### Motivation / Background

- We run ad-hoc scripts under `./script` via `bin/rails runner ./script/<file>`.
  - I think it's common to run files as temporary scripts when you want to avoid putting complex data patches into DB migrations.
- When debugging, we need accurate backtraces and query call locations, but `Rails.backtrace_cleaner` hides frames outside the fixed "application code" paths.
- As a result, useful frames from `./script` (and similar custom dirs) are removed.

*Problem Example*

```
# ./script/verbose_logs.rb
User.where(id: 1).load
```

```
# On main, frames under `./script` are cleaned, so their backtraces don't appear.
$ bundle exec rails runner ./script/verbose_logs.rb
  User Load (1.0ms)  SELECT "users".* FROM "users" WHERE "users"."id" = 1
```

```
# Backtrace includes gem internals but omits our script frames we actually want.
$ BACKTRACE=1 bundle exec rails runner ./script/verbose_logs.rb
  User Load (1.0ms)  SELECT "users".* FROM "users" WHERE "users"."id" = 1
  ↳ activerecord (7.2.3) lib/active_record/log_subscriber.rb:122:in 'ActiveRecord::LogSubscriber#log_query_source'
```

### Detail

I updated `Rails::BacktraceCleaner` it so that directories directly under the application root - except for vendor - are not cleaned, based on the advice in https://github.com/rails/rails/pull/56333#issuecomment-3635662172.

```
# This PR
$ bundle exec rails runner ./script/verbose_logs.rb
  User Load (1.1ms)  SELECT "users".* FROM "users" WHERE "users"."id" = 1
  ↳ script/verbose_logs.rb:1:in '<top (required)>'
```

### Additional information

Who benefits

- Teams that keep runnable utilities in user-defined dirs, or modular-monolith setups that store app code in `./packs`, can keep those paths visible in backtraces for debugging.
- `ActiveSupport::BacktraceCleaner#remove_filters!` isn't a viable alternative because it strips out the useful default filters.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

skipkayhil commented:
To me it seems like a better configuration point would be to allow customizing `Rails.backtrace_cleaner` instead of allowing the `app_dirs` to be configured. `Rails::BacktraceCleaner` is just a (private) `ActiveSupport::BacktraceCleaner` that adds a few things:

- `BACKTRACE` to disable
- Convert absolute paths -> relative to `Rails.root`
- template render filtering
- app directory filtering

alpaca-tc committed:
Do not clean directories directly under the application root with `Rails::BacktraceCleaner`

Improved `Rails.backtrace_cleaner` so that paths located directly under the application's root directory are no longer silenced.

Co-Authored-by: Jean Boussier <jean.boussier@gmail.com>

alpaca-tc commented:
@byroot Thanks for the advice!
I've updated the changes based on your suggestions. 😄 

In the original tests, the `bin/` directory was being cleaned, but with this PR it is no longer cleaned. Personally, I believe this behavior matches the expected intention, but it could also be viewed as a compatibility-affecting change.
Could you review this changes 🙏
